### ISSUE TITLE
FTP server lacks tests for RNTO and RNFR.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4930]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#4930|
|Type|defect|
|Created|2011-03-08 23:57:32Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-rnfr-tests-4930|

calling RNTO while being authenticated, without calling RNFR will generate an unhandled exceptions since self._fromName is only defined in RNFR.

Here is a patch that identifies the problem:

Please let me know if you agree with with fix and I could provide a full patch together with the required tests.

```
#!diff
=== modified file 'TwistedCore-10.2.0/twisted/protocols/ftp.py'
--- TwistedCore-10.2.0/twisted/protocols/ftp.py	2010-12-23 07:46:10 +0000
+++ TwistedCore-10.2.0/twisted/protocols/ftp.py	2011-03-08 23:54:19 +0000
@@ -752,6 +752,8 @@
                 return BAD_CMD_SEQ, "PASS required after USER"
 
         elif self.state == self.AUTHED:
+            if cmd == 'RNTO':
+                return BAD_CMD_SEQ, "RNFR required before RNTO"
             method = getattr(self, "ftp_" + cmd, None)
             if method is not None:
                 return method(*params)
```

Attachments:

* [4930-rnto-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cd8/cd836810b0408d1441d69783f75b30643f8b2d53/40574759b518379c29dfee21c7ec82304f84b2ab.diff) (2842 bytes) - added by adiroiban on 2012-06-13 23:08:14Z - 
* [4930-rnto-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cd8/cd836810b0408d1441d69783f75b30643f8b2d53/4029ed085644c26195f71f4cd01adb08d79fabcf.diff) (3769 bytes) - added by adiroiban on 2012-06-17 18:35:13Z - 
* [4930-rnto-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cd8/cd836810b0408d1441d69783f75b30643f8b2d53/8896a32c461e653d8615c555b7517451faea611d.diff) (3770 bytes) - added by adiroiban on 2012-06-17 18:42:00Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4930 4930
type__defect defect
reporter__adiroiban adiroiban
priority__low low
milestone__ 
branch__branches_ftp_rnfr_tests_4930 branches/ftp-rnfr-tests-4930
branch_author__spiv spiv
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1299628652000000 1299628652000000
changetime__1342343922000000 1342343922000000
version__None None
owner__spiv spiv
cc__itamarst cc__adiroiban cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

It looks like this was already implemented in latest Twisted version.

I was not able to find the test for server so I start adding a test for this case.

While working at this, I also found that there is no test for a proper RNFR and RNTO for the server. There are many tests for the client.

I have attached a patch containing the 2 tests.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @adiroiban|
|-|-|

Thanks. Before this can land the testcase names:

 1. should start with `test_`
 1. need a docstring

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

Hi,

I have attached a new diff.

Sorry for missing the docstring and the test_ prefix.

I have also renamed testSYST to test_SYST and added a docstring since I used it as an example for starting the new test.

For test_RNFRandRNTO I have also added an 'integration' check for file renaming on disk.

Please let me know if it needs other changes.
Cheers

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv set owner to spiv|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|spiv commented|
|-|-|

As the original goal of this ticket has been satisfied I've updated the summary to reflect the goal of the current patches.

```
            userHome=self.directory, 
```

It's a bit of a misfeature that the anonymous user implies read-only access, as it leads to ugliness like adding cruft about logging-in non-anonymous users to write tests for RNFR/RNTO.  The capabilities of the client should be orthogonal to how the login was done.  Oh well, something to worry about in another ticket.

```
 	371	    def test_RNFRandRNTO(self): 
 	372	        """ 
 	373	        Sending the RNFR command followed by RNTO should perform a 
 	374	        successful rename operation. 
 	375	        """ 
```

Say "will perform" rather than "should perform".  Be assertive in test docstrings :)

Also, the claim isn't strictly true: it also requires that the source and destination filenames are valid (or even more pedantically, that IFTPShell implementation accepts the `rename(src, dest)` call).  So I'd say "Sending the RNFR command followed by RNTO, with valid filenames, will perform a successful rename operation".

```
            ["250 Requested File Action Completed OK"], 
```

This test highlights that the Title Cased Phrase here is pretty weird and inconsistent with the other messages.  Not the fault of this patch, but if you're inclined to please submit a patch to tidy this up.

As my only comment requesting action is a tweak to a docstring, I'll go ahead and submit this (assuming the tests actually pass).

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|spiv commented|
|-|-|

(In [34797]) Branch to ftp-rnfr-tests-4930.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|spiv commented|
|-|-|

(In [34798]) Commit adiroiban's 4930-rnto-3.diff.

Refs [#4930](https://github.com/twisted/twisted/issues/4930).

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|spiv commented|
|-|-|

I forgot to add: we could do with even more tests here, e.g. exercising what happens when the server encounters an error when tries to perform the rename.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|spiv set status to `closed`|
|-|-|

(In [34801]) Merge ftp-rnfr-tests-4930: Add more FTP server tests.

Author: adiroiban
Reviewer: spiv
Fixes: [#4930](https://github.com/twisted/twisted/issues/4930)

Adds some test coverage for the FTP server implementation of the RNFR and RNTO
verbs.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

Hi spiv and many thanks for the review.

Your comments are very helpful.

Hope it the future the patches will be perfect :)

Cheers!
